### PR TITLE
Subject Group Viewer - Feb 2020 update

### PR DESCRIPTION
**Pull request by shaunanoordin**
## PR Overview

Package: `lib-classifier`
Project: Survos Survey adjacent

This PR attempts to fix two minor issues with the Subject Group Viewer.

- [x] Zoom range changed from 0.5-2x to 1-4x
  - Context: users mentioned it didn't make sense to be able to _zoom out_ on cells. Sounds legit.
- [x] The UI for focused and selected need to be re-thought.
  - Context: users have provided feedback indicating they're not sure if clicking on a selected cell also "unselects" it. (It does, but it may not be visually apparent.)
  - Analysis: this is a UI issue. The "you have focus on the cell" visuals (thick blue borders) can be confused with the "you have selected a cell" visuals (thin blue borders, and blue tint overlay), possibly more apparent on non-black-and-white photos.

### Testing

URL:
- https://local.zooniverse.org:8080/?env=staging&project=1900&workflow=3412 - original project using black-and-white cells. The blue overlay for "selected cells" is very visible here.
- (TBC) - other projects using photo (colour) images. The blue overlay is a lot LESS visible, which is an issue for determining if a cell is "selected"

Testing method:
- Open the Classifier on any of the links above. ☝️ 
- Click on any cell (we'll call this Cell 1) to **select** it.
- Cell 1 now has **focus** (THICK blue border) and is **selected** (blue overlay/tint)
- Click on another cell (Cell 2) to **select** it.
- Cell 2 now has **focus** (THICK blue border) and is **selected** (blue overlay)
- Cell 1 should now have _no focus_ but remain **selected** (thin blue border + blue overlay)
- Click on Cell 2 to to **unselect** it
- Cell 2 should now have **focus** (thick blue border) but is _not selected_ (no overlay)
- Compare Cell 1 and Cell 2 visually. Judge if they're visually distinct from one another.

### Status

WIP

### CODE REVIEW

**Comment by shaunanoordin** [comment]
## Design Discussion!

@beckyrother , I need to tag you in for your expertise. Here's the situation...

- Chris tried a version of the SURVOS Subject Group Viewer with cat photos instead of black-and-white cells like the original https://fe-project.preview.zooniverse.org/projects/darkeshard/survos-testing-2020/classify/workflow/3412?env=staging
- This lead to an observation: when there are _coloured photos,_ it's harder to determine whether a cell that _has focus_ is also _selected (checked)_ or _not._
  - (Nyeh, I don't have a demo link for this at the moment, sorry)

Anyway, currently we have this design:

Screen Shot 2021-02-19 at 22 41 49

- User clicks on the centre tile. The tile has _focus_ (thick blue border) and is now _selected_ (blue overlay, and thin blue border (hidden behind thick border))
- User clicks on the centre tile again. The tile continues to have _focus_ but is now _unselected_ (no blue overlay)

_(Extra clarification: "focus" as in "it has keyboard/navigation focus". "Selected" as in "the checkbox is ticked".)_

I think the (potential) confusion comes from the visual indicators. The cell border is pulling double duty to indicate "selected" and "has focus", so the meaning of the border isn't always clear. Normally we'd then be able to determine that a cell is "selected"/"not selected" by the secondary visual indicator of the blue tint... but the tint might not be so obvious on _colour photos._

### UI change proposal

So this is what I have in mind:

- Basically, we separate the duties of the cell border. One cell border isn't enough to visually indicate the condition of two state values ("has focus", "is selected")
- So let's have TWO cell borders! raaargh double the poweeer

Screen Shot 2021-02-19 at 22 49 39
_Centre cell has focus, but isn't selected/checked)_

Screen Shot 2021-02-19 at 22 49 50
_Centre cell has focus, and IS selected/checked)_

- I'm thinking of keeping the visual indicators for "is selected" (thin blue border, blue tint) the same
- and then I'll move the visual indicator for "has focus" into a _smaller white rectangle_ within the cell.

So, any thoughts on this? [Good? Bad? Ugly?](https://www.youtube.com/watch?v=aJCSNIl2Pls)

**Comment by shaunanoordin** [comment]
Bonus: current (not-PR update) visuals with a bunch of coloured photos

Screen Shot 2021-02-19 at 23 57 15

Note to self: if Becky asks for more of these examples, switch the image source with `xlinkHref={`https://picsum.photos/id/${['1021', '1022', '1023', '1024', '1025'][Math.floor(Math.random() * 5)]}/300/300`}`

**Comment by shaunanoordin** [comment]
I'm marking this PR ready for review! It's function complete, just waiting for notes from design.

**Review by eatyourgreens — COMMENTED** [review]
Code changes look good to me. Could the individual frame focus use the same focus ring that we use elsewhere? I think the styles are defined in the Grommet theme.

**Comment by beckyrother** [comment]
Hi @shaunanoordin ! I like the idea of two borders to differentiate between selected and focus. I worry the white border might not be "active" enough if you know what I mean? What about something like this green: 
https://projects.invisionapp.com/dsm/zooniverse/primary-brand/asset/colors/5bfeed65c541d90011cfc9a2 to really stand out?

**Comment by camallen** [comment]
@shaunanoordin what's the status of this PR? Chris has a use case for this in the next week.

**Comment by shaunanoordin** [comment]
@camallen Currently addressing feedback on colours, and trying to figure out why the tests are suddenly failing out of nowhere.

**Comment by eatyourgreens** [comment]
There's a long-running test, for one of the subject viewer zoom components, which intermittently times out and fails. That seems to be the problem here so I've re-run the tests. The failing test might be #1346.

**Comment by shaunanoordin** [comment]
## PR Update

- I have two options for additional feedback.
- Also, tests AREN'T failing locally, so wtf, CI Tests?

**Updated Focus Colours Option A:** Becky's #06FE76 recommendation

Screen Shot 2021-02-25 at 14 38 33

Screen Shot 2021-02-25 at 14 38 44

**Updated Focus Colours Option B:** standard focus ring colours (as seen when links and buttons have focus), aka accent-2, aka #ADDDE0

Screen Shot 2021-02-25 at 14 39 36

Screen Shot 2021-02-25 at 14 39 45

### Notes

When Jim recommended using a standard "focus ring", he meant "use the same light _outer outline_ as used when a link or button has focus"

See: 

image

```
outline: none;
box-shadow: 0 0 2px 2px #addde0;
```

The current limitation with the SubjectViewerGroup is that each cell is clipped to its specific area; if the cells weren't clipped, they'd "step" on each other and a method to dynamically manage the z-levels will be required. This limits my current ability to create an _outer outline_ easily, so the best I can manage with Option B is the colours.

### Status

I'm sticking with an inner outline, as initially proposed. Between Option A and Option B, I'm leaning strongly towards Option A since the high-saturation minty-green provides a lot of contrast on a wider array of graphics and photos. 👍

I think this PR is ready once I figure out whassup with those tests.

**Comment by shaunanoordin** [comment]
> There's a long-running test, for one of the subject viewer zoom components, which intermittently times out and fails. That seems to be the problem here so I've re-run the tests. The failing test might be #1346.

Goddangit! Thanks Jim, I just saw that 504 Gateway timeout error on the build. It failed again, so I'm re-running it once more.

**Comment by shaunanoordin** [comment]
Woot, tests pass! @eatyourgreens @beckyrother does the latest update look good to you? (To be clear, I went with Option A, which is Becky's minty green)

**Review by beckyrother — COMMENTED** [review]
Sorry but just so I'm understanding, Minty Green is #06FE76?

**Review by eatyourgreens — APPROVED** [review]
Looks good to me.
